File: 1137. N-th Tribonacci Number.java
import java.util.HashMap;

class Solution {
    HashMap<Integer,Integer> map = new HashMap<>();
    public int tribonacci(int n) {
        if(map.containsKey(n)){
            return map.get(n);
        }else if(n==0){
            return 0;
        }else if(n == 1 || n == 2){
            return 1;
        }
        map.put(n,tribonacci(n-1) + tribonacci(n-2) + tribonacci(n-3));
        return map.get(n);
    }
}

File: 1137. N-th Tribonacci Number.py
class Solution:
    def tribonacci(self, n: int) -> int:
        cache = dict()
        def dp(num):
            if num < 0 or num == 0:
                return 0
            elif num == 1 or num == 2:
                return 1
            elif num in cache:
                return cache[num]
            cache[num] = dp(num-1) + dp(num-2) + dp(num-3)
            return cache[num]
        return dp(n)